### PR TITLE
Fixed #34849 -- Avoided raising RuntimeWarning about import-time queries when apps are reinitialized with test tools.

### PR DESCRIPTION
**Pull request by shangxiao**
https://code.djangoproject.com/ticket/34849

### CODE REVIEW

File: tests/postgres_tests/test_apps.py [hunk 1]
```diff
@@ -27,6 +27,13 @@
 
 @unittest.skipUnless(connection.vendor == "postgresql", "PostgreSQL specific tests")
 class PostgresConfigTests(TestCase):
+    def test_install_app(self):
```
Comment (shangxiao) [inline tests/postgres_tests/test_apps.py:30]: We could potentially merge this test with the one below as they're both testing contrib.postgres' ready

**Comment by PoByBolek** [comment]
This feels weird. It just suppresses the issue by sidestepping the runtime warning, but doesn't fix the underlying problem. The tests still have the same isolation issue -- we just don't see it anymore.

**Comment by shangxiao** [comment]
> This feels weird. It just suppresses the issue by sidestepping the runtime warning, but doesn't fix the underlying problem. The tests still have the same isolation issue -- we just don't see it anymore.

Suggestions are welcome 😊

`contrib.postgres`' ready func either gets type information in a connection created signal, or _directly_ because connections already exist. I think this behaviour is acceptable, though others may disagree.

The other option is to have an "all apps ready signal", that may require larger discussion & thought as it would add yet more non-trivial behaviour to maintain 🤷‍♂️

If option 2 isn't desirable, then I think this is an acceptable compromise since `contrib.postgres` isn't really doing anything bad.

**Comment by PoByBolek** [comment]
> Suggestions are welcome 😊

Would if I could 😅 

I guess I would have explicitly silenced the warning using something like 
```python
with warnings.catch_warnings():
    warnings.filterwarnings("ignore", cursor.APPS_NOT_READY_WARNING_MSG, RuntimeWarning)
    cursor.execute(...)
```

so that we'd still be using the normal `CursorWraper` to execute the query.

**Comment by shangxiao** [comment]
Yes another possible option 👍

**Comment by felixxm** [comment]
> Yes another possible option 👍

We could also use `ignore_warnings()`, e.g.
```diff
diff --git a/django/contrib/postgres/signals.py b/django/contrib/postgres/signals.py
index a3816d3d30..ca6db0146f 100644
--- a/django/contrib/postgres/signals.py
+++ b/django/contrib/postgres/signals.py
@@ -3,13 +3,16 @@ import functools
 from django.db import connections
 from django.db.backends.base.base import NO_DB_ALIAS
 from django.db.backends.postgresql.psycopg_any import is_psycopg3
+from django.test.utils import ignore_warnings
 
 
 def get_type_oids(connection_alias, type_name):
     with connections[connection_alias].cursor() as cursor:
-        cursor.execute(
-            "SELECT oid, typarray FROM pg_type WHERE typname = %s", (type_name,)
-        )
+        msg = "Accessing the database during app initialization is discouraged."
+        with ignore_warnings(category=RuntimeWarning, message=msg):
+            cursor.execute(
+                "SELECT oid, typarray FROM pg_type WHERE typname = %s", (type_name,)
+            )
         oids = []
         array_oids = []
         for row in cursor:

```

**Comment by felixxm** [comment]
Personally, I'd prefer `ignore_warnings`. I don't think there is much we can do to avoid this.

@shangxiao Do you have time to keep working on this?

**Comment by shangxiao** [comment]
@felixxm Sure 👍

**Comment by charettes** [comment]
Could we simply not emit an a warning when `apps.stored_app_configs` is present instead?

This seems very much an artifact of using `override_settings(INSTALLED_APPS)` which results in `apps.set_installed_apps` being called. Silencing a false-positive warning (remember that it's meant to prevent module level queries) seems counter productive here.

```diff
diff --git a/django/db/backends/utils.py b/django/db/backends/utils.py
index 0aa94cc30b..604942dbb7 100644
--- a/django/db/backends/utils.py
+++ b/django/db/backends/utils.py
@@ -61,7 +61,7 @@ def callproc(self, procname, params=None, kparams=None):
                 "Keyword parameters for callproc are not supported on this "
                 "database backend."
             )
-        if not apps.ready:
+        if not apps.ready and not apps.stored_app_configs:
             warnings.warn(self.APPS_NOT_READY_WARNING_MSG, category=RuntimeWarning)
         self.db.validate_no_broken_transaction()
         with self.db.wrap_database_errors:
```

**Comment by shangxiao** [comment]
@charettes Sounds like a nice idea, worth asking though: does this mean you've changed your mind from https://code.djangoproject.com/ticket/33143#comment:2? :)

I agree with Adam that folks doing queries during ready is a more advanced step so folks probably know what they're doing.

The main issue I've seen as with others (eg also on IRC like Curtis mentioned) is people accidentally querying at module level.

**Comment by charettes** [comment]
I don't think the above recommendation goes against [my initial comment](https://code.djangoproject.com/ticket/33143#comment:2).

`Apps.set_installed_apps` can only successfully called when apps are already ready which means that there is not risk of performing module level queries at that point

https://github.com/django/django/blob/574ee4023e15cfb195833edfbaed353f8021c62f/django/apps/registry.py#L339C9-L357

`Apps.set_installed_apps` is the only method augmenting `stored_app_configs` and thus making it truthy

The sole purpose of `set_installed_apps` is to accommodate `@override_settings(INSTALLED_APPS)` which is a testing utility and has little do to with the intent of the warning?

**Comment by shangxiao** [comment]
Well that sounds fortuitous 👍.  One thing option I was thinking of suggesting was blocking only ORM queries and allowing raw but that still sounds like it has the potential to open up issues similar to that which you've described; getting db metadata is quite niche in this case.

File: django/db/backends/utils.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def callproc(self, procname, params=None, kparams=None):
                 "Keyword parameters for callproc are not supported on this "
                 "database backend."
             )
-        if not apps.ready:
+        if not apps.ready and not apps.stored_app_configs:
```
Comment (shangxiao) [inline django/db/backends/utils.py:64]: Do we want to store these in a property eg `Apps.is_initializing` where we can also put a comment/docstring that clarifies what Simon explained about `stored_app_configs` is mostly tied to `@override_settings` ? 🤔

Comment (felixxm) [reply django/db/backends/utils.py:64]: I don't think it's worth a new attribute.

> where we can also put a comment/docstring that clarifies what Simon explained about

Yes, please.

**Comment by felixxm** [comment]
@shangxiao Thanks :+1:

**Comment by felixxm** [comment]
I adjusted tests added in fbd16438f46bc2128926958ad24331da5d1b406f.

**Pull request merged**